src/Microsoft.ML.TensorFlow/TensorFlow/Tensor.cs

```diff
@@ -446,6 +446,104 @@ internal static unsafe TFTensor CreateString(byte[] buffer)
             return new TFTensor(handle);
         }
 
+        internal static unsafe byte[] DecodeString(TFTensor tensor)
+        {
+            if (tensor == null)
+                throw new ArgumentNullException(nameof(tensor));
+            //
+            // TF_STRING tensors are encoded with a table of 8-byte offsets followed by
+            // [offset1, offset2,...,offsetn, s1size, s1bytes, s2size, s2bytes,...,snsize,snbytes]
+            //
+            var src = TF_TensorData(tensor.handle);
+            using (var status = new TFStatus())
+            {
+                IntPtr dst = IntPtr.Zero;
+                UIntPtr dst_len = UIntPtr.Zero;
+                TF_StringDecode((sbyte*)(src + 8), tensor.TensorByteSize - 8, (sbyte**)&dst, &dst_len, status.handle);
+                var ok = status.StatusCode == TFCode.Ok;
+                if (!ok)
+                    return null;
+                var buffer = new byte[(int)dst_len];
+                Marshal.Copy(dst, buffer, 0, buffer.Length);
+                return buffer;
+            }
+        }
+
+        internal static unsafe TFTensor CreateStringTensor(byte[][] buffer, TFShape shape)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException(nameof(buffer));
```

Comment:
>ArgumentNullException [](start = 26, length = 21)

You know how we have Contracts.CheckValue and other things, right?
What happens is we mark exception and we can catch it later and if it's marked it can tell us something happened, but we anticipated via our checks.
If it's just naked exception, it's something we don't anticipated.
It was mostly case for command line, and tests, but I think it's still good practice to use it in our code.

Can you go through code and change all throw exception to `Contracts.Something()`

Reply:
Good catch! Actually this code was mostly taken from TensorFlowSharp and we never bother to adapt it to ML.NET standards. So, I will try to see if it possible to do it in this PR. Otherwise I will make it an issue and do it in another one.

---
In reply to: [256952363](https://github.com/dotnet/machinelearning/pull/2546#discussion_r256952363) [](ancestors = 256952363)

Reply:
Well technically we have Contracts.CheckValue(buffer) which will do check is  buffer empty or not, and throw exception. But this code is already improvement.

---
In reply to: [256964415](https://github.com/dotnet/machinelearning/pull/2546#discussion_r256964415) [](ancestors = 256964415,256952363)